utilities/download_old_firefox.py

```diff
@@ -7,45 +7,24 @@
 if __name__ == "__main__":
     base_url = "https://download-installer.cdn.mozilla.net"
     today = date.today()
+    download_month = None
     # Set correct month for URL builder
     if len(f"{today.month}") < 2:
-        current_month = f"0{today.month}"
+        download_month = f"0{today.month - 1}"
+    elif today.month == 1: #  january
+        download_month == 12
```

Comment:
I think yes... but no?
I'm going to run this script in 2 weeks on Jan 3, 2025, IIUC, it will:
1. try setting today's year (2025), see that the month is 1/January, 
2. try and download from a non-existent /nightly/2025/12/* folder.

I think if you're fudging a month back one, you need to similarly drop a year. Which might mean setting `download_month=today.month` somewhere and then in the special January case override the `download_month=download_month-1` or whatever you do in PyLand.

I'm getting lost in this thread w/ timedelta and relativedelta, but there might be some builtins for this https://stackoverflow.com/questions/9724906/python-date-of-the-previous-month

Reply:
Still not 100% happy with it, but think it handles a couple edges:

```py
from datetime import date

# today = date.today()

base_url = "https://download-installer.cdn.mozilla.net"

for i in range(1, 12):
    today = date.fromisoformat(f"2019-{i:02}-19")
    download_year = today.year

    download_month = f"{today.month - 1:02}"
    if today.month == 1:
        download_year = download_year - 1
        download_month = 12

    download_dir = f"{base_url}/pub/firefox/nightly/{download_year}/{download_month}/"

    print(download_dir)
```

<details>
<summary>OUTPUT</summary>

```
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2018/12/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/01/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/02/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/03/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/04/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/05/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/06/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/07/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/08/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/09/
https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/10/
```
</details>

Reply:
Closer&hellip;

```py
from datetime import date, timedelta

base_url = "https://download-installer.cdn.mozilla.net"

for i in range(1, 12):
    today = date.fromisoformat(f"2019-{i:02}-19")
    firefox_date = today - timedelta(weeks=4)
    firefox_date = firefox_date.replace(day=15)
    print(firefox_date)

    download_dir = f"{base_url}/pub/firefox/nightly/{firefox_date.year}/{firefox_date.month:02}/"

    print(download_dir)
```

DATE | URL
:----|:----
2018-12-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2018/12/
2019-01-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/01/
2019-02-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/02/
2019-03-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/03/
2019-04-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/04/
2019-05-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/05/
2019-06-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/06/
2019-07-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/07/
2019-08-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/08/
2019-09-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/09/
2019-10-15 | https://download-installer.cdn.mozilla.net/pub/firefox/nightly/2019/10/

Reply:
Oh I didn't know about `timedelta`, that should be the solution